datahub/interaction/test/admin_csv_import/test_row_form.py

```
@@ -776,6 +777,134 @@ def test_cleaned_data_as_serializer_dict_for_service_delivery(self):
+        assert interaction.event == event
+        assert interaction.notes == data['notes']
+        assert interaction.created_by == user
+        assert interaction.modified_by == user
+        assert interaction.source == source
+
+        assert list(interaction.contacts.all()) == [contact]
+        assert interaction.dit_participants.count() == 2
+
+        actual_advisers_and_teams = Counter(
+            (dit_participant.adviser, dit_participant.team)
+            for dit_participant in interaction.dit_participants.all()
+        )
+        expected_advisers_and_teams = Counter(
+            (
+                (adviser_1, adviser_1.dit_team),
+                (adviser_2, adviser_2.dit_team),
+            ),
+        )
+
+        assert actual_advisers_and_teams == expected_advisers_and_teams
+
+    def test_save_rolls_back_on_error(self, monkeypatch):
+        """Test that save() rolls back if there's an error."""
+        monkeypatch.setattr(
+            'datahub.interaction.admin_csv_import.row_form.InteractionDITParticipant',
+            Mock(side_effect=ValueError),
+        )
+
+        user = AdviserFactory()
+        adviser = AdviserFactory(first_name='Neptune', last_name='Doris')
+        contact = ContactFactory(email='unique@company.com')
+        service = random_service()
+        communication_channel = random_communication_channel()
+        source = {'test-source': 'test-value'}
+
+        data = {
+            'kind': Interaction.KINDS.interaction,
+            'date': '02/03/2018',
+            'adviser_1': adviser.name,
+            'contact_email': contact.email,
+            'service': service.name,
+            'communication_channel': communication_channel.name,
+        }
+        form = InteractionCSVRowForm(data=data)
+
+        assert form.is_valid()
+
+        with pytest.raises(ValueError):
+            form.save(user, source=source)
+
+        assert not Interaction.objects.exists()
+
```

It would be good to have a test for when a contact is not matched. My understanding is that `is_valid` would silently return but the save would fail?

---

I've added a test and made it a little bit more robust for that situation.